paho/client_options.go

```diff
@@ -0,0 +1,148 @@
+// the 'server' string is in the go style of "<server>:<port>"
+func OpenTCP6Conn(server string) func(*Client) error {
+	return func(c *Client) error {
+		conn, err := net.Dial("tcp6", server)
+		if err != nil {
+			return err
+		}
+		c.Conn = conn
+		return nil
+	}
+}
+
+// OpenTLSConn will try and create a TCP connection to be used by the client,
+// and use the TLS config from 'config' to negotiate a secure connection,
+// the 'server' string is in the go style of "<server>:<port>"
+func OpenTLSConn(server string, config *tls.Config) func(*Client) error {
+	return func(c *Client) error {
+		conn, err := tls.Dial("tcp", server, config)
+		if err != nil {
+			return err
+		}
+		c.Conn = conn
+		return nil
+	}
+}
+
+// OpenTLS6Conn will try and create a TCP6 connection to be used by the client,
+// and use the TLS config from 'config' to negotiate a secure connection,
+// the 'server' string is in the go style of "<server>:<port>"
+func OpenTLS6Conn(server string, config *tls.Config) func(*Client) error {
+	return func(c *Client) error {
+		conn, err := tls.Dial("tcp6", server, config)
+		if err != nil {
+			return err
+		}
+		c.Conn = conn
+		return nil
+	}
+}
+
+// WithPinger will set the client's internal ping mechanism to be the Pinger
+// passed into this function, if this is not used the default Pinger provided
+// in this library will be used.
+func WithPinger(p Pinger) func(*Client) error {
+	return func(c *Client) error {
+		c.PingHandler = p
+		return nil
+	}
+}
+
+// WithRouter will set the client's internal message router to be the Router
+// passed into this function, if this is not used the default Router provided
+// in this library will be used.
+func WithRouter(r Router) func(*Client) error {
+	return func(c *Client) error {
+		c.Router = r
+		return nil
+	}
+}
+
+// DefaultMessageHandler sets the client's internal message routing to always
+// invoke the MessageHandler passed to this function when the client receives
+// a message. This cannot be combined with other Routers, the last function
+// to set a Router will be the configuration used by the client.
+func DefaultMessageHandler(h MessageHandler) func(*Client) error {
+	return func(c *Client) error {
+		c.Router = &SingleHandlerRouter{
+			messageHandler: h,
+		}
+		return nil
+	}
+}
+
+// WithMIDService sets the client's internal message ID service to be the
+// MIDService passed into this function. If this option is not used the default
+// MIDService provided by this library will be used.
+func WithMIDService(m MIDService) func(*Client) error {
+	return func(c *Client) error {
+		c.MIDs = m
+		c.MIDs.Clear()
+		return nil
+	}
+}
+
+// WithPersistence sets the client's internal message persistence mechanism
+// to be the Persistence passed into this function. If this option is not used
+// *no* persistence mechanism is used. This library currently also provides a
+// memory based Persistence.
+func WithPersistence(p Persistence) func(*Client) error {
```

Comment:
This new version is minimalistic and I really like it, do you think that it should implement message persistence? As I see it'll be needed to write a higher-level abstraction per each project for features like: reconnecting, lazy connection, messages buffering, etc.

Reply:
Thanks, the persistence mechanisms are pluggable, I haven't looked to provide a proper one yet as they can be a bit of a pain to debug, although if you were interested in doing one, either as a PR to the main code or a standalone library that others can use, that'd be cool

Reply:
Some features like autoreconnect, message buffering etc from the current v3 client don't seem to actually make it any simpler for the application writer, and add more code for me to maintain. Happy to look at any such features and see if their is value in them and how they might be implemented in the v5 client

Reply:
@alsm Forgive me if I am not understanding the reason for this design; nevertheless I see room for improvement.

Function values are not comparable. As a consequence, functions that return functions cannot be tested in a data-driven way (there's no declarative way to ensure the output is correct; you have to call the function and test its behavior). I'm seeing them liberally used here, and this style of code is not typical in Go. Granted, the functions seem straightforward. Their role is to configure the client. But why not let that be a data declaration? Why does it have to involve higher-order functions? The fields being initialized are public anyway.

A Go motto is to accept interfaces and return structs. A function is more like an interface than a struct; it can only be called.